subsys/bluetooth/controller/ll_sw/isoal.c

```diff
@@ -0,0 +1,409 @@
+
+	if (sdu_complete) {
+		err = sink->session.sdu_emit(sink, sdu);
+	}
+
+	return err;
+}
+
+static isoal_status_t isoal_rx_append_to_sdu(struct isoal_sink *sink,
+					     const struct isoal_pdu_rx *pdu_meta,
+					     bool is_end_fragment)
+{
+	isoal_status_t err = ISOAL_STATUS_OK;
+	const uint8_t *pdu_payload          = pdu_meta->pdu->cis.payload;
+	/* Note length field in same location for both CIS and BIS */
+	isoal_pdu_len_t packet_available    = pdu_meta->pdu->cis.length;
+
+	/* TODO error handling and lost packet */
+
+	/* While there is something left of the packet to consume */
+	while (packet_available > 0) {
+		const isoal_status_t err_alloc = isoal_rx_allocate_sdu(sink, pdu_meta);
+		struct isoal_sdu_produced *sdu = &sink->sdu_production.sdu;
+		uint8_t *sdu_buf = &sdu->contents.codewords[sink->sdu_production.sdu_written];
+
+		err |= err_alloc;
+
+		/*
+		 * For this SDU we can only consume of packet, bounded by:
+		 *   - What can fit in the destination SDU.
+		 *   - What remains of the packet.
+		 */
+		const size_t consume_len = MIN(
+			packet_available,
+			sink->sdu_production.sdu_available
+		);
+
+		LL_ASSERT(sdu_buf);
+		LL_ASSERT(pdu_payload);
+		if (pdu_meta->meta->status == ISOAL_PDU_STATUS_VALID) {
+			memcpy(sdu_buf, pdu_payload + pdu_meta->meta->read_offset, consume_len);
+		}
+		sdu_buf     += consume_len;
+		pdu_payload += consume_len;
+		sink->sdu_production.sdu_written   += consume_len;
+		sink->sdu_production.sdu_available -= consume_len;
+		packet_available                   -= consume_len;
+
+		sdu->status |= pdu_meta->meta->status;
+
+		bool force_emit = (packet_available == 0) && is_end_fragment;
+
+		const isoal_status_t err_emit = isoal_rx_try_emit_sdu(sink, force_emit);
+
+		err |= err_emit;
+	}
+
+	LL_ASSERT(packet_available == 0);
+	return err;
+}
+
+
+/**
+ * @brief Consume a PDU: Copy contents into SDU(s) and emit to a sink
+ * @details Destination sink may have an already partially built SDU
+ *
+ * @param sink[in,out]    Destination sink with bookkeeping state
+ * @param pdu_meta[out]  PDU with meta information (origin, timing, status)
+ *
+ * @return TODO
+ */
+static isoal_status_t isoal_rx_packet_consume(struct isoal_sink *sink,
+					      const struct isoal_pdu_rx *pdu_meta)
+{
+	isoal_pdu_cnt_t id, prev_id;
+
+	isoal_status_t err = ISOAL_STATUS_OK;
+
+	/*
+	 *   SDU of length N can be partially filled, over a number of cases:
+	 *      0                          N-1
+	 *   1) [#########                   ] : We have the beginning. Not yet completed.
+	 *   2) [#########                   ] : We lack the end. Completed. Frame is lost.
+	 *   3) [                   #########] : We lack the beginning. Frame lost
+	 *   4) [         #########          ] : We lack the beginning. Not yet completed, sdu lost
+	 *   5) [#########          #########] : We lack something. Completed. Frame is lost.
+	 *   6) [                            ] : Predict
```

Comment:
Considering a simplified case of 1 SDU := 3x PDU, we get the cases below.
Where V := Valid, A:= Absent, . := Not received yet.
(Brackets delimit PDUs)

```
$ echo {[VVVV],[AAAA],[....]}{[VVVV],[AAAA],[....]}{[VVVV],[AAAA],[....]} | linear | grep -Ev "\.+.*[AV]" | nl
     1  [VVVV][VVVV][VVVV] 
     2  [VVVV][VVVV][AAAA] 
     3  [VVVV][VVVV][....] 
     4  [VVVV][AAAA][VVVV] 
     5  [VVVV][AAAA][AAAA] 
     6  [VVVV][AAAA][....] 
     7  [VVVV][....][....] 
     8  [AAAA][VVVV][VVVV] 
     9  [AAAA][VVVV][AAAA] 
    10  [AAAA][VVVV][....] 
    11  [AAAA][AAAA][VVVV] 
    12  [AAAA][AAAA][AAAA] 
    13  [AAAA][AAAA][....] 
    14  [AAAA][....][....] 
    15  [....][....][....] 
```

Reply:
will add this when implementing error handling